### PR TITLE
Issue #12164: Enable Checker Framework

### PR DESCRIPTION
**Pull request by nrmancuso**
#12164.

Activates all Checkers as warn

### CODE REVIEW

**Comment by romani** [comment]
Error: [ERROR] error: OutOfMemoryError (max memory = 1820327936, total memory = 1820327936, free memory = 644792)

Looks like we need profiles to split execution. Or find what Checker is prone to oom

**Comment by nrmancuso** [comment]
>Looks like we need profiles to split execution. Or find what Checker is prone to oom

I am working on splitting them up now

File: pom.xml [hunk 1]
```diff
@@ -383,6 +385,11 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+      <groupId>org.checkerframework</groupId>
+      <artifactId>checker-qual</artifactId>
+      <version>3.25.0</version>
```
Comment (nrmancuso) [inline pom.xml:391]: Need to use var here.

File: pom.xml [hunk 2]
```diff
@@ -2382,6 +2389,452 @@
...
+                <arg>-Awarns</arg>
+                <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED</arg>
+                <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED</arg>
+                <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED</arg>
+                <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED</arg>
+                <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED</arg>
+                <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED</arg>
+                <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED</arg>
+                <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED</arg>
+                <arg>-J--add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED</arg>
```
Comment (nrmancuso) [inline pom.xml:2430]: These exports are required for many checkers, and compilation was failing without them.

File: pom.xml [hunk 3]
```diff
@@ -2382,6 +2389,452 @@
...
+                </path>
+              </annotationProcessorPaths>
+              <annotationProcessors>
+                <annotationProcessor>org.checkerframework.checker.nullness.NullnessChecker</annotationProcessor>
+                <annotationProcessor>org.checkerframework.checker.optional.OptionalChecker</annotationProcessor>
+                <annotationProcessor>org.checkerframework.checker.interning.InterningChecker</annotationProcessor>
+              </annotationProcessors>
+              <compilerArgs>
+                <!-- -Awarns turns type-checking errors into warnings. -->
+                <arg>-Awarns</arg>
```
Comment (nrmancuso) [inline pom.xml:2421]: We only warn, we do not fail compilation

File: pom.xml [hunk 4]
```diff
@@ -2382,6 +2389,452 @@
...
+                <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED</arg>
+                <arg>-J--add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED</arg>
+              </compilerArgs>
+            </configuration>
+          </plugin>
+        </plugins>
+      </build>
+      <properties>
+        <!-- Needed for animal-sniffer-maven-plugin version 1.19 (version 1.20 is fixed). -->
+        <animal.sniffer.skip>true</animal.sniffer.skip>
```
Comment (nrmancuso) [inline pom.xml:2438]: This plugin builda signatures of APIs and checks classes against previously generated signatures.

**Review by nrmancuso — COMMENTED** [review]
Comments:

File: .github/workflows/checker-framework.yml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+          - checker-framework-regex-property-key-compiler-message
+          - checker-framework-purity-value-returns
+
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up JDK
+        uses: actions/setup-java@v3
+        with:
+          distribution: 'zulu'
+          java-version: 17
```
Comment (nrmancuso) [inline .github/workflows/checker-framework.yml:29]: Java 11 in CI was showing inconsistent behavior for `-Awarns`, but fine on my local. Trying Java 17 now to see if this helps.

File: pom.xml [hunk 6]
```diff
@@ -2382,6 +2389,454 @@
...
+            <artifactId>maven-compiler-plugin</artifactId>
+            <version>${maven.compiler.plugin.version}</version>
+            <configuration>
+              <!-- Must fork or else JVM arguments are ignored. -->
+              <fork>true</fork>
+              <compilerArguments>
+                <Xmaxerrs>10000</Xmaxerrs>
+                <Xmaxwarns>10000</Xmaxwarns>
+                <meminitial>1024m</meminitial>
+                <maxmem>4048m</maxmem>
```
Comment (nrmancuso) [inline pom.xml:2574]: MAVEN_OPTS is not respected by forked processes.

File: pom.xml (hunk 2 above)
Comment (Vyom-Yadav) [reply pom.xml:2430]: You can add jvm arguments in https://github.com/checkstyle/checkstyle/blob/master/.mvn/jvm.config

**Comment by nrmancuso** [comment]
Closed in favor of https://github.com/checkstyle/checkstyle/pull/12194

**Pull request closed without merging**